## Gathering data

In [1]:
import pandas as pd
import numpy as np
import requests
import json
import tweepy


#### Reading the twitter archive enhanced csv file.

In [2]:
twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')

twitter_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

#### Reading the image predicitons tsv file using python's requests library.

In [3]:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)

localfile = url.split('/')[-1]
with open(localfile, mode='wb') as file:
    file.write(response.content)

In [4]:
image_predictions = pd.read_csv('image-predictions.tsv', sep='\t')

image_predictions.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

#### Reading the twitter data using Twitter API (tweepy)

In [5]:
consumer_key = 'J0UThDBO9siolq9pfZmcssywF'
consumer_secret = 'Gl1Bn8rBO7ePznuUluN5Y7RnHRLuWfnrculT8sLu6knPDjrp7f'
access_token = '1155634178811146245-P7ONVnV63xjqt6iXC2N2yi3jO836VL'
access_secret = 'Ws32qLAgaHITsAhuaRQ3JN6yDxNaLfTMvCdzybRSZWcCN'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True)

In [6]:
tweet_ids = twitter_archive.tweet_id

tweet_ids.head()

0    892420643555336193
1    892177421306343426
2    891815181378084864
3    891689557279858688
4    891327558926688256
Name: tweet_id, dtype: int64

source: https://stackoverflow.com/questions/34379168/how-to-save-the-tweets-in-json-to-txt-file-in-python-3-4

In [7]:
Exception_list = []
with open('tweet_json.txt', 'w', encoding='utf-8') as file:
    for tweet_id in tweet_ids:
        try:
            tweet = api.get_status(tweet_id, tweet_mode='extended')        
            json.dump(tweet._json, file)
            file.write('\n')
        except Exception as e:
            Exception_list.append(tweet_id)
            print(e)

[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status found with that ID.'}]
[{'code': 144, 'message': 'No status fou

In [11]:
tweet_json = pd.read_json('tweet_json.txt', orient='records', lines=True )

In [9]:
tweet_json.columns

Index(['contributors', 'coordinates', 'created_at', 'display_text_range',
       'entities', 'extended_entities', 'favorite_count', 'favorited',
       'full_text', 'geo', 'id', 'id_str', 'in_reply_to_screen_name',
       'in_reply_to_status_id', 'in_reply_to_status_id_str',
       'in_reply_to_user_id', 'in_reply_to_user_id_str', 'is_quote_status',
       'lang', 'place', 'possibly_sensitive', 'possibly_sensitive_appealable',
       'quoted_status', 'quoted_status_id', 'quoted_status_id_str',
       'quoted_status_permalink', 'retweet_count', 'retweeted',
       'retweeted_status', 'source', 'truncated', 'user'],
      dtype='object')

In [12]:
tweet_json.head()

contributors  coordinates          created_at display_text_range  \
0           NaN          NaN 2017-08-01 16:23:56            [0, 85]   
1           NaN          NaN 2017-08-01 00:17:27           [0, 138]   
2           NaN          NaN 2017-07-31 00:18:03           [0, 121]   
3           NaN          NaN 2017-07-30 15:58:51            [0, 79]   
4           NaN          NaN 2017-07-29 16:00:24           [0, 138]   

                                            entities  \
0  {'hashtags': [], 'symbols': [], 'user_mentions...   
1  {'hashtags': [], 'symbols': [], 'user_mentions...   
2  {'hashtags': [], 'symbols': [], 'user_mentions...   
3  {'hashtags': [], 'symbols': [], 'user_mentions...   
4  {'hashtags': [{'text': 'BarkWeek', 'indices': ...   

                                   extended_entities  favorite_count  \
0  {'media': [{'id': 892420639486877696, 'id_str'...           37273   
1  {'media': [{'id': 892177413194625024, 'id_str'...           32049   
2  {'media': [{'id': 891815175371796480, 'id_str'...           24148   
3  {'media': [{'id': 891689552724799489, 'id_str'...           40589   
4  {'media': [{'id': 891327551943041024, 'id_str'...           38826   

   favorited                                          full_text  geo  ...  \
0      False  This is Phineas. He's a mystical boy. Only eve...  NaN  ...   
1      False  This is Tilly. She's just checking pup on you....  NaN  ...   
2      False  This is Archie. He is a rare Norwegian Pouncin...  NaN  ...   
3      False  This is Darla. She commenced a snooze mid meal...  NaN  ...   
4      False  This is Franklin. He would like you to stop ca...  NaN  ...   

   quoted_status  quoted_status_id quoted_status_id_str  \
0            NaN               NaN                  NaN   
1            NaN               NaN                  NaN   
2            NaN               NaN                  NaN   
3            NaN               NaN                  NaN   
4            NaN               NaN                  NaN   

   quoted_status_permalink  retweet_count  retweeted  retweeted_status  \
0                      NaN           7984      False               NaN   
1                      NaN           5930      False               NaN   
2                      NaN           3917      False               NaN   
3                      NaN           8142      False               NaN   
4                      NaN           8816      False               NaN   

                                              source truncated  \
0  <a href="http://twitter.com/download/iphone" r...     False   
1  <a href="http://twitter.com/download/iphone" r...     False   
2  <a href="http://twitter.com/download/iphone" r...     False   
3  <a href="http://twitter.com/download/iphone" r...     False   
4  <a href="http://twitter.com/download/iphone" r...     False   

                                                user  
0  {'id': 4196983835, 'id_str': '4196983835', 'na...  
1  {'id': 4196983835, 'id_str': '4196983835', 'na...  
2  {'id': 4196983835, 'id_str': '4196983835', 'na...  
3  {'id': 4196983835, 'id_str': '4196983835', 'na...  
4  {'id': 4196983835, 'id_str': '4196983835', 'na...  

[5 rows x 32 columns]

In [19]:
tweet_df = tweet_json[['id', 'retweet_count', 'favorite_count', 'retweeted', 'display_text_range']]

tweet_df = tweet_df.rename(columns={'id':'tweet_id'})

tweet_df.head()

tweet_id  retweet_count  favorite_count  retweeted  \
0  892420643555336193           7984           37273      False   
1  892177421306343426           5930           32049      False   
2  891815181378084864           3917           24148      False   
3  891689557279858688           8142           40589      False   
4  891327558926688256           8816           38826      False   

  display_text_range  
0            [0, 85]  
1           [0, 138]  
2           [0, 121]  
3            [0, 79]  
4           [0, 138]

## Assessing data

In [20]:
twitter_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [21]:
twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [22]:
twitter_archive.duplicated().sum()

0

There are no duplicate values in this dataset.

In [23]:
twitter_archive.isnull().sum()

tweet_id                         0
in_reply_to_status_id         2278
in_reply_to_user_id           2278
timestamp                        0
source                           0
text                             0
retweeted_status_id           2175
retweeted_status_user_id      2175
retweeted_status_timestamp    2175
expanded_urls                   59
rating_numerator                 0
rating_denominator               0
name                             0
doggo                            0
floofer                          0
pupper                           0
puppo                            0
dtype: int64

In [24]:
twitter_archive.in_reply_to_status_id.head(35)

0              NaN
1              NaN
2              NaN
3              NaN
4              NaN
5              NaN
6              NaN
7              NaN
8              NaN
9              NaN
10             NaN
11             NaN
12             NaN
13             NaN
14             NaN
15             NaN
16             NaN
17             NaN
18             NaN
19             NaN
20             NaN
21             NaN
22             NaN
23             NaN
24             NaN
25             NaN
26             NaN
27             NaN
28             NaN
29             NaN
30    8.862664e+17
31             NaN
32             NaN
33             NaN
34             NaN
Name: in_reply_to_status_id, dtype: float64

In [25]:
twitter_archive.in_reply_to_user_id.head(35)

0              NaN
1              NaN
2              NaN
3              NaN
4              NaN
5              NaN
6              NaN
7              NaN
8              NaN
9              NaN
10             NaN
11             NaN
12             NaN
13             NaN
14             NaN
15             NaN
16             NaN
17             NaN
18             NaN
19             NaN
20             NaN
21             NaN
22             NaN
23             NaN
24             NaN
25             NaN
26             NaN
27             NaN
28             NaN
29             NaN
30    2.281182e+09
31             NaN
32             NaN
33             NaN
34             NaN
Name: in_reply_to_user_id, dtype: float64

In [26]:
twitter_archive.timestamp.dtype

dtype('O')

In [27]:
twitter_archive.text.sample()

688    This is Frank. He wears sunglasses and walks h...
Name: text, dtype: object

In [28]:
twitter_archive.doggo.value_counts()

None     2259
doggo      97
Name: doggo, dtype: int64

In [29]:
twitter_archive.floofer.value_counts()

None       2346
floofer      10
Name: floofer, dtype: int64

In [30]:
twitter_archive.pupper.value_counts()

None      2099
pupper     257
Name: pupper, dtype: int64

In [31]:
twitter_archive.puppo.value_counts()

None     2326
puppo      30
Name: puppo, dtype: int64

In [32]:
test = twitter_archive[twitter_archive.tweet_id.isin(image_predictions.tweet_id)]

len(test)

2075

This means there are several tweet_ids not present in image_prediction data.

In [80]:
twitter_archive.retweeted_status_id.notnull().sum()

181

This means there are 181 retweeted entries in this dataframe which should not be there.

In [33]:
image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [34]:
image_predictions.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [35]:
image_predictions.img_num.value_counts()

1    1780
2     198
3      66
4      31
Name: img_num, dtype: int64

In [36]:
image_predictions.tweet_id.duplicated().sum()

0

So there are no duplicate tweet ids in this dataframe.

In [37]:
image_predictions.jpg_url.duplicated().sum()

66

In [38]:
image_predictions.jpg_url.value_counts().head(15)

https://pbs.twimg.com/media/CzG425nWgAAnP7P.jpg    2
https://pbs.twimg.com/media/CxqsX-8XUAAEvjD.jpg    2
https://pbs.twimg.com/media/Ct72q9jWcAAhlnw.jpg    2
https://pbs.twimg.com/media/Cs_DYr1XEAA54Pu.jpg    2
https://pbs.twimg.com/media/CiyHLocU4AI2pJu.jpg    2
https://pbs.twimg.com/media/CvoBPWRWgAA4het.jpg    2
https://pbs.twimg.com/media/CrXhIqBW8AA6Bse.jpg    2
https://pbs.twimg.com/media/Cbs3DOAXIAAp3Bd.jpg    2
https://pbs.twimg.com/media/Ct2qO5PXEAE6eB0.jpg    2
https://pbs.twimg.com/media/C2kzTGxWEAEOpPL.jpg    2
https://pbs.twimg.com/media/CkNjahBXAAQ2kWo.jpg    2
https://pbs.twimg.com/media/CiibOMzUYAA9Mxz.jpg    2
https://pbs.twimg.com/media/CWyD2HGUYAQ1Xa7.jpg    2
https://pbs.twimg.com/media/CsrjryzWgAAZY00.jpg    2
https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg    2
Name: jpg_url, dtype: int64

All the jpg_urls having counts > 1 are duplicated/retweeted. Checking one of them to find the difference between both the entries.

In [39]:
image_predictions.query('jpg_url == "https://pbs.twimg.com/media/Cp6db4-XYAAMmqL.jpg"')

tweet_id                                          jpg_url  \
1380  765222098633691136  https://pbs.twimg.com/media/Cp6db4-XYAAMmqL.jpg   
1427  772615324260794368  https://pbs.twimg.com/media/Cp6db4-XYAAMmqL.jpg   

      img_num         p1   p1_conf  p1_dog       p2   p2_conf  p2_dog  \
1380        1  dalmatian  0.556595    True  whippet  0.151047    True   
1427        1  dalmatian  0.556595    True  whippet  0.151047    True   

                                  p3   p3_conf  p3_dog  
1380  American_Staffordshire_terrier  0.096435    True  
1427  American_Staffordshire_terrier  0.096435    True

In [40]:
twitter_archive.query('tweet_id == "765222098633691136"')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
854  765222098633691136                    NaN                  NaN   

                     timestamp  \
854  2016-08-15 16:22:20 +0000   

                                                source  \
854  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
854  This is Gromit. He's pupset because there's no...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
854                       NaN                        NaN   

                                         expanded_urls  rating_numerator  \
854  https://twitter.com/dog_rates/status/765222098...                10   

     rating_denominator    name doggo floofer pupper puppo  
854                  10  Gromit  None    None   None  None

In [41]:
twitter_archive.query('tweet_id == "772615324260794368"')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
800  772615324260794368                    NaN                  NaN   

                     timestamp  \
800  2016-09-05 02:00:22 +0000   

                                                source  \
800  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
800  RT @dog_rates: This is Gromit. He's pupset bec...         7.652221e+17   

     retweeted_status_user_id retweeted_status_timestamp  \
800              4.196984e+09  2016-08-15 16:22:20 +0000   

                                         expanded_urls  rating_numerator  \
800  https://twitter.com/dog_rates/status/765222098...                10   

     rating_denominator    name doggo floofer pupper puppo  
800                  10  Gromit  None    None   None  None

This dataframe contains tweets that were retweeted

In [42]:
image_predictions.query('p1_dog == False and p2_dog == False and p3_dog == False')

tweet_id                                            jpg_url  \
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
45    666786068205871104    https://pbs.twimg.com/media/CUDmZIkWcAAIPPe.jpg   
50    666837028449972224    https://pbs.twimg.com/media/CUEUva1WsAA2jPb.jpg   
51    666983947667116034    https://pbs.twimg.com/media/CUGaXDhW4AY9JUH.jpg   
53    667012601033924608    https://pbs.twimg.com/media/CUG0bC0U8AAw2su.jpg   
56    667065535570550784    https://pbs.twimg.com/media/CUHkkJpXIAA2w3n.jpg   
69    667188689915760640    https://pbs.twimg.com/media/CUJUk2iWUAAVtOv.jpg   
73    667369227918143488    https://pbs.twimg.com/media/CUL4xR9UkAEdlJ6.jpg   
77    667437278097252352    https://pbs.twimg.com/media/CUM2qWaWoAUZ06L.jpg   
78    667443425659232256    https://pbs.twimg.com/media/CUM8QZwW4AAVsBl.jpg   
93    667549055577362432    https://pbs.twimg.com/media/CUOcVCwWsAERUKY.jpg   
94    667550882905632768    https://pbs.twimg.com/media/CUObvUJVEAAnYPF.jpg   
96    667724302356258817    https://pbs.twimg.com/media/CUQ7tv3W4AA3KlI.jpg   
98    667766675769573376    https://pbs.twimg.com/media/CURiQMnUAAAPT2M.jpg   
100   667782464991965184    https://pbs.twimg.com/media/CURwm3cUkAARcO6.jpg   
106   667866724293877760    https://pbs.twimg.com/media/CUS9PlUWwAANeAD.jpg   
107   667873844930215936    https://pbs.twimg.com/media/CUTDtyGXIAARxus.jpg   
112   667911425562669056    https://pbs.twimg.com/media/CUTl5m1WUAAabZG.jpg   
115   667937095915278337    https://pbs.twimg.com/media/CUT9PuQWwAABQv7.jpg   
117   668142349051129856    https://pbs.twimg.com/media/CUW37BzWsAAlJlN.jpg   
118   668154635664932864    https://pbs.twimg.com/media/CUXDGR2WcAAUQKz.jpg   
123   668226093875376128    https://pbs.twimg.com/media/CUYEFlQXAAUkPGm.jpg   
130   668291999406125056    https://pbs.twimg.com/media/CUZABzGW4AE5F0k.jpg   
132   668466899341221888    https://pbs.twimg.com/media/CUbfGbbWoAApZth.jpg   
140   668544745690562560    https://pbs.twimg.com/media/CUcl5jeWsAA6ufS.jpg   
...                  ...                                                ...   
1839  837482249356513284    https://pbs.twimg.com/media/C59VqMUXEAAzldG.jpg   
1844  838916489579200512    https://pbs.twimg.com/media/C6RkiQZUsAAM4R4.jpg   
1847  839290600511926273    https://pbs.twimg.com/media/C6XBt9XXEAEEW9U.jpg   
1851  840370681858686976    https://pbs.twimg.com/media/C6mYrK0UwAANhep.jpg   
1853  840696689258311684    https://pbs.twimg.com/media/C6rBLenU0AAr8MN.jpg   
1869  844580511645339650    https://pbs.twimg.com/media/C7iNfq1W0AAcbsR.jpg   
1886  847962785489326080    https://pbs.twimg.com/media/C8SRpHNUIAARB3j.jpg   
1887  847971574464610304    https://pbs.twimg.com/media/C8SZH1EWAAAIRRF.jpg   
1891  849051919805034497    https://pbs.twimg.com/media/C8hwNxbXYAAwyVG.jpg   
1892  849336543269576704    https://pbs.twimg.com/media/C8lzFC4XcAAQxB4.jpg   
1900  851464819735769094    https://pbs.twimg.com/media/C9ECujZXsAAPCSM.jpg   
1902  851861385021730816    https://pbs.twimg.com/media/C8W6sY_W0AEmttW.jpg   
1905  852226086759018497  https://pbs.twimg.com/ext_tw_video_thumb/85222...   
1906  852311364735569921    https://pbs.twimg.com/media/C9QEqZ7XYAIR7fS.jpg   
1910  853299958564483072    https://pbs.twimg.com/media/C9eHyF7XgAAOxPM.jpg   
1931  859074603037188101    https://pbs.twimg.com/media/C-wLyufW0AA546I.jpg   
1936  860184849394610176    https://pbs.twimg.com/media/C-_9jWWUwAAnwkd.jpg   
1937  860276583193509888    https://pbs.twimg.com/media/C_BQ_NlVwAAgYGD.jpg   
1940  860924035999428608    https://pbs.twimg.com/media/C_KVJjDXsA

All of these columns have p1_dog, p2_dog & p3_dog values False which means the images in this entries aren't of dogs.

In [43]:
tweet_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2333 entries, 0 to 2332
Data columns (total 5 columns):
tweet_id              2333 non-null int64
retweet_count         2333 non-null int64
favorite_count        2333 non-null int64
retweeted             2333 non-null bool
display_text_range    2333 non-null object
dtypes: bool(1), int64(3), object(1)
memory usage: 75.3+ KB


In [44]:
tweet_df.retweeted.value_counts()

False    2333
Name: retweeted, dtype: int64

None of the tweets were retweeted. This column is not needed in this dataframe.

### Summary

#### Quality Issues

**twitter archive dataframe**
1. Incorrect datatypes for in_reply_to_status_id, in_reply_to_user_id columns, timestamp fields & tweet_id.
2. Unneccessary html tags in some of the values in source column.
3. Text column contains untruncated text instead of displayable text.
4. Many of the tweet ids present in twitter_archive df are not present in image_predictions df.
5. There are retweeted entries in this dataframe.

**image prediction dataframe**
1. Incorrect datatypes for p1, p2 & p3 & tweet_id
2. There should be space instead of '_' in p1,p2 & p3 columns.
3. The names in p1, p2 & p3 are not consistent, sometimes start with uppercase and other times with lowercase.
4. There are several entries with no dog images in the tweets.

**tweet df**
1. There is no need to have retweeted column in this dataframe since all of its values are False.
2. Incorrect datatype for tweet_id column.

#### Tidiness Issues

1. Retweet count & Favorite count should be merged from tweet_df into twitter_archive dataframe.
2. Merge the doggo, floffer, pupper & puppo columns of twitter_archive into a single column dog_stage.
3. There is no need to have retweeted columns (retweeted_status_id, retweeted_status_user_id &   retweeted_status_timestamp)
4. Merge p1, p2 & p3 into a single prediction column with highest confidence.
5. Merge all the relevant columns and create one master clean dataframe.

## Cleaning data

In [143]:
twitter_archive_clean = twitter_archive.copy()

**twitter archive dataframe**

1. Incorrect datatypes for in_reply_to_status_id, in_reply_to_user_id columns, timestamp fields & tweet_id.
   

#### Define

 - change in_reply_to_status_id & in_reply_to_user_id from float to int
 - timestamp from object to datetime
 - tweet_id from int to string

#### Code

In [154]:
twitter_archive_clean.in_reply_to_status_id = twitter_archive_clean.in_reply_to_status_id.fillna(0)
twitter_archive_clean.in_reply_to_user_id = twitter_archive_clean.in_reply_to_user_id.fillna(0)

twitter_archive_clean.in_reply_to_status_id = twitter_archive_clean.in_reply_to_status_id.astype(int)
twitter_archive_clean.in_reply_to_user_id = twitter_archive_clean.in_reply_to_user_id.astype(int)

twitter_archive_clean.timestamp = pd.to_datetime(twitter_archive_clean.timestamp)
twitter_archive_clean.retweeted_status_timestamp = pd.to_datetime(twitter_archive_clean.retweeted_status_timestamp)

twitter_archive_clean.tweet_id = twitter_archive_clean.tweet_id.astype(str)

tweet_df.tweet_id = tweet_df.tweet_id.astype(str)

#### Test

In [54]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null object
in_reply_to_status_id         2356 non-null int64
in_reply_to_user_id           2356 non-null int64
timestamp                     2356 non-null datetime64[ns, UTC]
source                        2356 non-null category
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null datetime64[ns, UTC]
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dty

2. Unneccessary html tags in some of the values in source column.

#### Define

Remove html tags and just leave the text in the source column and then convert it into category datatype

#### Code

In [155]:
twitter_archive_clean.source = twitter_archive_clean.source.str.replace(r'<(?:a\b[^>]*>|/a>)', '')
twitter_archive_clean.source = twitter_archive_clean.source.astype('category')

#### Test

In [56]:
twitter_archive_clean.source.sample()

280    Twitter for iPhone
Name: source, dtype: category
Categories (4, object): [TweetDeck, Twitter Web Client, Twitter for iPhone, Vine - Make a Scene]

In [57]:
twitter_archive_clean.source.value_counts()

Twitter for iPhone     2221
Vine - Make a Scene      91
Twitter Web Client       33
TweetDeck                11
Name: source, dtype: int64

3. Text column contains untruncated text instead of displayable text.


#### Define

Extract the displayable text from tweet_df and use it in twitter_archive_clean dataframe

#### Code

In [68]:
print(twitter_archive_clean.text[1])

print(twitter_archive_clean.text[2])

This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV
This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB


In [156]:
twitter_archive_clean = pd.merge(twitter_archive_clean, tweet_df[['tweet_id', 'display_text_range']], on='tweet_id')

In [157]:
for i, j in twitter_archive_clean.iterrows():
    text_range = j.display_text_range
    display = j.text[text_range[0]:text_range[1]]
    twitter_archive_clean.set_value(i, 'text', display)
    

/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:4: FutureWarning: set_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead
  after removing the cwd from sys.path.


#### Test

In [74]:
print(twitter_archive_clean.text[1])

print(twitter_archive_clean.text[2])

This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10
This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10


4. Many of the tweet ids present in twitter_archive are not present in image_predictions.


#### Define

Keep only the entries of tweet_ids present in both twitter_archive dataframe and image_predictions dataframe.

#### Code

In [145]:
twitter_archive_clean = twitter_archive_clean[twitter_archive_clean.tweet_id.isin(image_predictions.tweet_id)]



#### Test

In [147]:
len(twitter_archive_clean[~twitter_archive_clean.tweet_id.isin(image_predictions.tweet_id)])

0

5. There are retweeted entries in this dataframe

#### Define

The retweeted_status_id column value should be null in all the cases since there should be no retweeted entries in this df. The current value is 81.

In [150]:
twitter_archive_clean.retweeted_status_id.notnull().sum()

81

#### Code

In [151]:
twitter_archive_clean = twitter_archive_clean[twitter_archive_clean.retweeted_status_id.isnull()]

#### Test

In [152]:
twitter_archive_clean.retweeted_status_id.notnull().sum()

0

In [153]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1994 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      1994 non-null int64
in_reply_to_status_id         23 non-null float64
in_reply_to_user_id           23 non-null float64
timestamp                     1994 non-null object
source                        1994 non-null object
text                          1994 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 1994 non-null object
rating_numerator              1994 non-null int64
rating_denominator            1994 non-null int64
name                          1994 non-null object
doggo                         1994 non-null object
floofer                       1994 non-null object
pupper                        1994 non-null object
puppo                         1994 non-null object
dtypes: float64(4), int64(3), object(1

**image predictions dataframe**

In [226]:
image_predictions_clean = image_predictions.copy()

1. Incorrect datatypes for p1, p2 & p3 & tweet_id


#### Define

Change the datatypes for p1,p2 & p3 from string to category and tweet_id datatype from int to str.


#### Code

In [227]:
image_predictions_clean.p1 = image_predictions_clean.p1.astype('category')

image_predictions_clean.p2 = image_predictions_clean.p2.astype('category')

image_predictions_clean.p3 = image_predictions_clean.p3.astype('category')

image_predictions_clean.tweet_id = image_predictions_clean.tweet_id.astype(str)

#### Test

In [165]:
image_predictions_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null object
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null category
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null category
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null category
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), category(3), float64(3), int64(1), object(2)
memory usage: 174.9+ KB


2. There should be space instead of '_' in p1,p2 & p3 columns.


#### Define

The '_' should be replaced with space from the values in p1, p2 & p3 columns.

#### Code

In [228]:
image_predictions_clean = image_predictions_clean.replace('_', ' ', regex=True)

#### Test

In [175]:
image_predictions_clean.head(5)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh springer spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature pinscher   
2        1         German shepherd  0.596461    True            malinois   
3        1     Rhodesian ridgeback  0.408143    True             redbone   
4        1      miniature pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland sheepdog  0.061428    True  
1  0.074192    True  Rhodesian ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

3. The names in p1, p2 & p3 are not consistent, sometimes start with uppercase and other times with lowercase.

#### Define

All the names in columns p1, p2 and p3 should start with lower case letters.

#### Code

In [229]:
image_predictions_clean.p1 = image_predictions_clean.p1.str.lower()

image_predictions_clean.p2 = image_predictions_clean.p2.str.lower()

image_predictions_clean.p3 = image_predictions_clean.p3.str.lower()

#### Test

In [187]:
image_predictions_clean.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  welsh springer spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature pinscher   
2        1         german shepherd  0.596461    True            malinois   
3        1     rhodesian ridgeback  0.408143    True             redbone   
4        1      miniature pinscher  0.560311    True          rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    shetland sheepdog  0.061428    True  
1  0.074192    True  rhodesian ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature pinscher  0.222752    True  
4  0.243682    True             doberman  0.154629    True

**tweet_df**

In [230]:
tweet_df_clean = tweet_df.copy()

1. There is no need to have retweeted column in this dataframe since all of its values are False.

#### Define

Delete retweeted column from the dataframe

#### Code

In [231]:
tweet_df_clean.head()

tweet_id  retweet_count  favorite_count  retweeted  \
0  892420643555336193           7984           37273      False   
1  892177421306343426           5930           32049      False   
2  891815181378084864           3917           24148      False   
3  891689557279858688           8142           40589      False   
4  891327558926688256           8816           38826      False   

  display_text_range  
0            [0, 85]  
1           [0, 138]  
2           [0, 121]  
3            [0, 79]  
4           [0, 138]

In [233]:
tweet_df_clean = tweet_df_clean.drop('retweeted', axis=1)

#### Test

In [234]:
tweet_df_clean.head()

tweet_id  retweet_count  favorite_count display_text_range
0  892420643555336193           7984           37273            [0, 85]
1  892177421306343426           5930           32049           [0, 138]
2  891815181378084864           3917           24148           [0, 121]
3  891689557279858688           8142           40589            [0, 79]
4  891327558926688256           8816           38826           [0, 138]

2. Incorrect datatype for tweet_id column.

#### Define

Change tweet_id column datatype from int to str.

#### Code

In [235]:
tweet_df_clean.tweet_id = tweet_df_clean.tweet_id.astype(str)

#### Test

In [236]:
tweet_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2333 entries, 0 to 2332
Data columns (total 4 columns):
tweet_id              2333 non-null object
retweet_count         2333 non-null int64
favorite_count        2333 non-null int64
display_text_range    2333 non-null object
dtypes: int64(2), object(2)
memory usage: 73.0+ KB
